[clang] Diagnose dangling issues for the "Container<GSLPointer>" case.

hokein opened the pull request:
This pull request enhances the GSL lifetime analysis to detect situations where a dangling `Container<GSLPointer>` object is constructed: 

```cpp
std::vector<std::string_view> bad = {std::string()}; // dangling
```

 The assignment case is not yet supported, but they will be addressed in a follow-up.

Fixes #100526 (excluding the `push_back` case).

hokein committed:
[clang] Diagnose dangling issues for "Container<GSLPointer>" case.

We teach the GSL lifetime analysis to handle this code path, particuly
when constructing the container<GSLPointer> object from a GSLOwner.

Xazax-hun reviewed (approved):
LGTM! Thanks!

usx95 commented on clang/lib/Sema/CheckExprLifetime.cpp:
```diff
@@ -363,10 +363,14 @@ static bool implicitObjectParamIsLifetimeBound(const FunctionDecl *FD) {
   return isNormalAsisgnmentOperator(FD);
 }
 
+bool isFirstTemplateArgumentGSLPointer(const TemplateArgumentList &TAs) {
+  return TAs.size() > 0 && TAs[0].getKind() == TemplateArgument::Type &&
+         isRecordWithAttr<PointerAttr>(TAs[0].getAsType());
+}
+
```
The motivation for checking this for first template args might not be immediately clear without seeing the usage.
I would suggest moving the `CTSD->hasAttr<OwnerAttr>())`  to this function and have a function `isContainerOfPointer(ClassTemplateSpecializationDecl*)` instead. Here we can document in this function that we are checking for "Owner<Pointer, ...>"

hokein replied:
Done.

usx95 commented on clang/lib/Sema/CheckExprLifetime.cpp:
```diff
@@ -470,10 +474,24 @@ static void visitFunctionCallArguments(IndirectLocalPath &Path, Expr *Call,
+                CTSD->hasAttr<OwnerAttr>()) {
+              VisitGSLPointerArg(Ctor->getConstructor()->getParamDecl(0),
+                                 Args[0], true);
+              return;
```
why `return` ? We don't return for ctor of GSLPointer case below. I think we might miss running lifetime bound analysis for rest of the args

usx95 commented on clang/lib/Sema/CheckExprLifetime.cpp:
```diff
@@ -470,10 +474,24 @@ static void visitFunctionCallArguments(IndirectLocalPath &Path, Expr *Call,
                            !Callee->getReturnType()->isReferenceType());
       } else {
-        if (auto *CCE = dyn_cast<CXXConstructExpr>(Call);
-            CCE && CCE->getConstructor()->getParent()->hasAttr<PointerAttr>())
-          VisitGSLPointerArg(CCE->getConstructor()->getParamDecl(0), Args[0],
-                             true);
+        if (auto *Ctor = dyn_cast<CXXConstructExpr>(Call)) {
+          const auto *ClassD = Ctor->getConstructor()->getParent();
+          // Constructing the Container<GSLPointer> case (e.g.
+          // std::optional<string_view>) case.
+          if (const auto *CTSD =
+                  dyn_cast<ClassTemplateSpecializationDecl>(ClassD)) {
+            if (isFirstTemplateArgumentGSLPointer(CTSD->getTemplateArgs()) &&
+                CTSD->hasAttr<OwnerAttr>()) {
+              VisitGSLPointerArg(Ctor->getConstructor()->getParamDecl(0),
+                                 Args[0], true);
+              return;
+            }
+          }
+          // Constructing the GSLPointer (e.g. std::string_view) case.
+          if (ClassD->hasAttr<PointerAttr>())
+            VisitGSLPointerArg(Ctor->getConstructor()->getParamDecl(0), Args[0],
+                               true);
+        }
```
This GSL analysis is done only on the first arg.
Can we separate this out and move outside the loop ?
```cpp
// Perform GSL analysis for the first argument
bool PerformGSLAnalysis = EnableLifetimeWarnings && !CheckCoroCall && !Callee->getParamDecl(0)->hasAttr<LifetimeBoundAttr>()
if (PerformGSLAnalysis) {
...
```

hokein replied:
The logic here is tightly coupled with the lifetimebound attribute. If the first argument is already visited due to the lifetimebound attribute, we avoid performing the GSL analysis on it again to prevent duplication. Therefore, both checks need to remain within the same loop.

usx95 replied:
Isn't `bool PerformGSLAnalysisForFirstArg = EnableLifetimeWarnings && !CheckCoroCall && !Callee->getParamDecl(0)->hasAttr<LifetimeBoundAttr>()` the sufficient condition ?

usx95 commented on clang/test/Sema/warn-lifetime-analysis-nocfg.cpp:
```diff
@@ -499,3 +518,28 @@ std::string_view test2(int i, std::optional<std::string_view> a) {
+
+namespace GH100526 {
+void test() {
+  std::vector<std::string_view> t1 = {std::string()}; // expected-warning {{object backing the pointer will be destroyed at the end}}
```
Can you try adding more than one args to init list for `std::vector`.
Eg: `std::vector<std::string_view> s = {std::string_view(),  std::string()}`

Does it work atm ?

hokein replied:
Yeah, it works. Added this case.

usx95 replied:
Great. Thanks.

usx95 reviewed (commented):
Thanks. This looks quite useful.

Since this is not limited to a hardcoded set of types/stl-containers,  this is a substantial change in the behaviour of general `Owner<Pointer> types` and warrants extra documentation both in GSL analysis docs and ReleaseNotes.

hokein commented:
> Since this is not limited to a hardcoded set of types/stl-containers, this is a substantial change in the behaviour of general Owner<Pointer> types and warrants extra documentation both in GSL analysis docs and ReleaseNotes.

Done.

hokein commented:
After further testing, this change introduces a new false positive -- we emit a return-stack-address warning for the following case, this is not correct.

```
std::vector<std::string_view> kk() {
  std::vector<std::string_view> s;
  return s;
}
```

Looking now.

usx95 commented on clang/docs/ReleaseNotes.rst:
```suggestion
- Clang now diagnoses cases where a dangling ``GSLOwner<GSLPointer>`` object is constructed, e.g. ``std::vector<string_view> v = {std::string()}`` (#GH100526).
```

hokein replied:
Done.

usx95 reviewed (commented):
Thanks. LG. Some comments on adding more tests.

hokein committed:
add more tests per comments.

hokein commented on clang/test/Sema/warn-lifetime-analysis-nocfg.cpp:
```diff
@@ -525,3 +544,31 @@ void test() {
 }
 } // namespace GH100549
+
+namespace GH100526 {
```
Done.

hokein commented on clang/test/Sema/warn-lifetime-analysis-nocfg.cpp:
Yeah, right. I added these two case.

hokein commented on clang/test/Sema/warn-lifetime-analysis-nocfg.cpp:
```diff
@@ -525,3 +544,31 @@ void test() {
+void test() {
+  std::vector<std::string_view> v1({std::string()}); // expected-warning {{object backing the pointer will be destroyed at the end}}
+  std::vector<std::string_view> v2({std::string(), std::string_view()}); // expected-warning {{object backing the pointer will be destroyed at the end}}
+  std::vector<std::string_view> v3({std::string_view(), std::string()}); // expected-warning {{object backing the pointer will be destroyed at the end}}
```
Good idea. Done.

hokein commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -6690,6 +6690,20 @@ When the Owner's lifetime ends, it will consider the Pointer to be dangling.
   }
 
+If a template class is annotated with [[gsl::Owner]], and the first instantiated
+template argument is a [[gsl::Pointer]] type, the analysis will consider the
```
Done.

hokein commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -6690,6 +6690,20 @@ When the Owner's lifetime ends, it will consider the Pointer to be dangling.
+.. code-block:: c++
+
+   int f() {
+     std::vector<std::string_view> v = {std::string();}; // v holds a dangling pointer.
```
Done.

hokein commented:
I have fixed the false positives, and added testcases for them. (I have tested this patch on our internal codebase, no new false positives being found).

usx95 reviewed (approved):
LGTM. Thanks.

hokein committed:
address review comments.

hokein commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -292,7 +292,7 @@ Improvements to Clang's diagnostics
 - Clang now warns for u8 character literals used in C23 with ``-Wpre-c23-compat`` instead of ``-Wpre-c++17-compat``.
 
-- Clang now diagnoses cases where a dangling `GSLOwner<GSLPointer>`` object is constructed, e.g. `std::vector<string_view> v = {std::string}` (#GH100526).
+- Clang now diagnoses cases where a dangling ``GSLOwner<GSLPointer>`` object is constructed, e.g. `std::vector<string_view> v = {std::string()};` (#GH100526).
```
Done.

hokein commented on clang/lib/Sema/CheckExprLifetime.cpp:
```diff
@@ -990,13 +1009,16 @@ static void checkExprLifetimeImpl(Sema &SemaRef,
-        IsLocalGslOwner = isRecordWithAttr<OwnerAttr>(L->getType());
+        IsLocalGslOwner =
+            isRecordWithAttr<OwnerAttr>(L->getType()) &&
+            !isContainerOfPointer(L->getType()->getAsRecordDecl());
```
Done.

aeubanks commented:
this looks like it has false positives: https://crbug.com/366074092

can we revert in the meantime?

hokein commented:
Thanks, it looks like this introduces a new false positive:

```
#include <optional>
std::optional<int*> func(int a) {
  if (a)
   return std::make_optional(nullptr); // emit a dangling.
}
```

I'm going to revert it.
